python/ray/util/client/server/dataservicer.py

```diff
@@ -22,6 +26,32 @@
 QUEUE_JOIN_SECONDS = 10
 
 
+def _get_reconnecting_from_context(context: Any) -> bool:
+    """
+    Get `reconnecting` from gRPC metadata, or False if not present
+    """
+    metadata = {k: v for k, v in context.invocation_metadata()}
+    val = metadata.get("reconnecting")
+    if val is None or val not in ("True", "False"):
+        logger.error(
+            f'Client connecting with invalid value for "reconnecting": {val}')
+        context.set_code(grpc.StatusCode.FAILED_PRECONDITION)
+        return None
+    return val == "True"
+
+
+def _should_cache(req: ray_client_pb2.DataRequest) -> bool:
+    """
+    Returns True if the response should to the given request should be cached,
+    false otherwise. At the moment the only requests we do not cache are
+    asynchronous gets, acks, and cleanup requests
```

Comment:
why we don't cache async gets?

Reply:
Mainly because they arrive out-of-order (all other requests get responses in order). Should be safe, since the gets won't modify state. Will update comment.